ipfs-cluster/trigger-server/triggerserver.go

```diff
@@ -8,88 +8,133 @@ import (
+	Secret string
+	Port int
+	RunnerPath string
+	Runner string
+	RunnerArgs string
+}
 
-// Secret is set by main
-var secret = ""
+const configPath = "/etc/trigger-server.json"
+var config TriggerConfig
 
 // IPFS Shell
 var sh = shell.NewShell("localhost:5001")
 
 func run(rw http.ResponseWriter, req *http.Request) {
-    if req.Method == "POST" {
-	if req.Header.Get("X-Auth") != secret {
-		log.Println("User failed authentication")
-		fmt.Fprintf(rw, "Invalid authentication\n")
-	} else {
-		// This command takes a while, so we'll hang out here while it runs
-		cmd := exec.Command("bash", "runner.sh", runnerNumNodes, runnerNumPins)
-		cmd.Dir = runnerPath
-		out, err := cmd.CombinedOutput()
-		// Add the output to IPFS
-		addr, ipfsErr := sh.Add(bytes.NewReader(out))
-		// Make it an IPFS link
-		addr = "/ipfs/" + addr
-		// Send the link to the client
-		fmt.Fprintf(rw, addr)
-		// Handle errors that may have occurred
+    if req.Method != "POST" {
+        fmt.Fprintf(rw, `Kubernetes-IPFS Trigger Server
+	https://github.com/ipfs/kubernetes-ipfs/
+	`)
+	return
+    }
+    if req.Header.Get("X-Auth") != config.Secret {
+	    log.Println("User failed authentication")
+	    rw.WriteHeader(http.StatusUnauthorized)
+	    fmt.Fprintf(rw, "Invalid authentication\n")
+	    return
+    }
+    // This command takes a while, so we'll hang out here while it runs
+    cmd := exec.Command("bash", "-c", fmt.Sprintf("%s%s %s", config.RunnerPath, config.Runner, config.RunnerArgs))
+    cmd.Dir = config.RunnerPath
+    pipeReader, pipeWriter := io.Pipe()
+    out := bytes.NewBufferString("")
+    cmd.Stdout = pipeWriter
+    cmd.Stderr = pipeWriter
+    go writeOutput(rw, out, pipeReader)
+    if err := cmd.Run(); err != nil {
+	    log.Printf("Failed to start runner (%s/%s %s): %s", config.RunnerPath, config.Runner, config.RunnerArgs, err)
+	    return
+    }
+    pipeWriter.Close()
+    log.Println("Finished execution")
+    // Add the output to IPFS
+    addr, ipfsErr := sh.Add(bytes.NewReader(out.Bytes()))
+    // Make it an IPFS link
+    addr = "/ipfs/" + addr
+    // Send the link to the client
+    fmt.Fprintf(rw, addr)
+    // Handle errors that may have occurred
+    if ipfsErr != nil {
+	    log.Println("IPFS Error adding output")
+	    log.Println(ipfsErr)
+    }
+}
+
+func writeOutput(res http.ResponseWriter, out *bytes.Buffer, pipeReader *io.PipeReader) {
+	buffer := make([]byte, 4096)
+	for {
+		n, err := pipeReader.Read(buffer)
 		if err != nil {
-			log.Println("Error from runner")
-			log.Println(err)
+			pipeReader.Close()
+			break
 		}
-		if ipfsErr != nil {
-			log.Println("IPFS Error adding output")
-			log.Println(ipfsErr)
+
+		data := buffer[0:n]
+		res.Write(data)
+		out.Write(data)
+		if f, ok := res.(http.Flusher); ok {
+			f.Flush()
 		}
+		//reset buffer
+		for i := 0; i < n; i++ {
+			buffer[i] = 0
+		}
+	}
+}
+
+func init() {
+    rand.Seed(time.Now().UnixNano())
+    configFile, readErr := ioutil.ReadFile(configPath)
```

Comment:
configuration handling to ->`main()`, but do create functions for it.. `DefaultConfig()`, `ReadConfig()`, `WriteConfig()`.. they can live in `config.go`